ggshield/core/oauth.py

```diff
@@ -363,27 +363,49 @@ def do_GET(self_) -> None:
                 callback_url: str = self_.path
                 parsed_url = urlparse.urlparse(callback_url)
                 if parsed_url.path == "/":
-                    try:
-                        self.oauth_client.process_callback(callback_url)
-                    except OAuthError as error:
-                        self_._end_request()
-                        # attach error message to the handler wrapper instance
-                        self.error_message = error.message
+                    server_error = self_._get_error_param(parsed_url)
+                    if server_error is not None:
+                        self_._end_request(200)
+                        self.error_message = server_error
                     else:
-                        self_._end_request(
-                            urljoin(self.oauth_client.dashboard_url, "authenticated"),
-                        )
-
-                    # indicate to the serve to stop
+                        try:
+                            self.oauth_client.process_callback(callback_url)
+                        except OAuthError as error:
+                            self_._end_request(400)
+                            # attach error message to the handler wrapper instance
+                            self.error_message = error.message
+                        else:
+                            self_._end_request(
+                                301,
+                                urljoin(
+                                    self.oauth_client.dashboard_url, "authenticated"
+                                ),
+                            )
+
+                    # indicate to the server to stop
                     self.complete = True
+                else:
+                    self_._end_request(404)
 
-            def _end_request(self_, redirect_url: Optional[str] = None) -> None:
+            def _get_error_param(
+                self_, parsed_url: urlparse.ParseResult
+            ) -> Optional[str]:
+                """
+                extract the value of the 'error' url param. If not present, return None.
+                """
+                params = urlparse.parse_qs(parsed_url.query)
+                if "error" in params:
+                    return params["error"][0]
+                return None
```

Comment:
You cannot do integration test due to the local server, but you could unit test functions like this to avoid simple mistakes.